mat_dp_pipeline/pipeline/pipeline.py

```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterator, overload
+
+import pandas as pd
+
+from mat_dp_pipeline.pipeline.calculation import ProcessedOutput, calculate
+from mat_dp_pipeline.pipeline.flatten_hierarchy import flatten_hierarchy
+from mat_dp_pipeline.pipeline.sparse_to_processable_input import to_processable_input
+from mat_dp_pipeline.sdf import StandardDataFormat, Year
+
+
+@dataclass(frozen=True)
+class LabelledOutput(ProcessedOutput):
+    year: Year
+    path: Path
+
+
+class PipelineOutput:
+    """The processed data from the pipeline in an easily accessible form."""
+
+    _by_year: dict[Year, dict[Path, LabelledOutput]]
+    _by_path: dict[Path, dict[Year, LabelledOutput]]
+    _length: int
+    _indicators: set[str]
+    _tech_metadata: pd.DataFrame
+
+    def __init__(self, data: list[LabelledOutput], tech_metadata: pd.DataFrame):
+        self._by_year = defaultdict(dict)
+        self._by_path = defaultdict(dict)
+        self._tech_metadata = tech_metadata
+
+        if data:
+            # We know from the computation that each LabelledOutput has the same set of indicators
+            self._indicators = data[0].indicators
+        else:
+            self._indicators = set()
+
+        assert all(
+            datum.indicators == self._indicators for datum in data
+        ), "Every single output must have the same set of indicators!"
+        assert all(
+            len(group["Material Unit"].unique()) == 1
+            and len(group["Production Unit"].unique()) == 1
+            for _, group in self._tech_metadata.groupby("Category")
+        ), "Every tech category must have unique matererial and production units!"
+
+        for output in data:
+            self._by_year[output.year][output.path] = output
+            self._by_path[output.path][output.year] = output
+        self._length = len(data)
+
+    def emissions(self, key: Path | str, indicator: str) -> pd.DataFrame:
+        """TODO:
+
+        Args:
+            indicator (str): _description_
```

Comment:
docstring issue

Reply:
Not exactly an issue - TODO. Left it for later.